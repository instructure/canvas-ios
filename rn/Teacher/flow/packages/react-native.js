//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
