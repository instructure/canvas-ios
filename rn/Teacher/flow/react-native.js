//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow
/* eslint-disable */

// define basic style prop types
// the internal ones are the best we have
import {
  type ____ViewStyleProp_Internal,
  type ____TextStyleProp_Internal,
  type ____ImageStyleProp_Internal,
} from 'react-native/Libraries/StyleSheet/StyleSheetTypes'

declare type ViewStyleProp = ____ViewStyleProp_Internal
declare type TextStyleProp = ____TextStyleProp_Internal
declare type ImageStyleProp = ____ImageStyleProp_Internal
