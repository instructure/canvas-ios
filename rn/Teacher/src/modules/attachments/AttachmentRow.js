//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import {
  StyleSheet,
  TouchableHighlight,
  Image,
  View,
  ActionSheetIOS,
  Alert,
} from 'react-native'
import Row from '../../common/components/rows/Row'
import images from '../../images'
import colors from '../../common/colors'
import { Circle } from 'react-native-progress'
import bytes from '../../utils/locale-bytes'
import i18n from 'format-message'
import { type Progress } from '../../canvas-api'

export type Props = {
  completed: boolean,
  title: string,
  progress: Progress,
  error: ?string,
  testID: string,
  onRemovePressed: () => any,
  onPress: () => any,
  onRetry: () => any,
  onCancel: () => any,
}

export default class AttachmentRow extends Component<Props, any> {
  render () {
    return (
      <Row
        title={this.props.title}
        subtitle={this.renderSubtitle()}
        renderImage={this.renderImage}
        accessories={this.removeButton()}
        onPress={this.props.onPress}
        testID={this.props.testID}
        accessible={false}
      />
    )
  }

  renderSubtitle () {
    if (this.props.error) return null
    const { loaded, total } = this.props.progress
    if (total > 0) {
      if (loaded >= total) {
        return bytes(total, { style: 'integer' })
      }
      return i18n('Uploading {loaded} of {total}', {
        loaded: bytes(loaded, { style: 'integer' }),
        total: bytes(total, { style: 'integer' }),
      })
    }

    return null
  }

  renderImage = () => {
    if (this.props.error) {
      return (
        <TouchableHighlight
          testID={`${this.props.testID}.icon.error`}
          underlayColor='transparent'
          onPress={this.onPressError}
          style={style.image}
          hitSlop={{ top: 12, right: 12, bottom: 12, left: 12 }}
          accessibilityLabel={i18n('Upload error')}
          accessibilityTraits='button'
        >
          <Image
            source={images.attachments.error}
            style={style.image}
          />
        </TouchableHighlight>
      )
    }

    if (this.props.completed) {
      return (
        <View
          style={style.image}
          accessibilityLabel={i18n('Upload complete')}
          accessible={true}
        >
          <Image
            source={images.attachments.complete}
            testID={`${this.props.testID}.icon.complete`}
          />
        </View>
      )
    }

    return (
      <TouchableHighlight
        testID={`${this.props.testID}.icon.progress`}
        underlayColor='transparent'
        onPress={this.props.onCancel}
        style={style.image}
        hitSlop={{ top: 12, right: 12, bottom: 12, left: 12 }}
        accessibilityLabel={i18n('Upload in progress')}
        accessibilityTraits='button'
      >
        <View style={{ flex: 1, justifyContent: 'center' }}>
          <View style={style.cancel}>
            <View
              style={[style.cancelInner, { backgroundColor: colors.primaryBrandColor }]}
            >
            </View>
          </View>
          <Circle
            size={24}
            borderWidth={0}
            thickness={2}
            progress={this.props.progress.loaded / this.props.progress.total}
            unfilledColor={'#F5F5F5'}
            borderColor={colors.primaryBrandColor}
            color={colors.primaryBrandColor}
          />
        </View>
      </TouchableHighlight>
    )
  }

  removeButton = () => {
    return (
      <TouchableHighlight
        onPress={this.onPressRemove}
        underlayColor='white'
        hitSlop={{ top: 8, bottom: 8, left: 8, right: 8 }}
        testID={`${this.props.testID}.remove.btn`}
        accessibilityLabel={i18n('Remove attachment')}
        accessibilityTraits='button'
      >
        <Image source={images.x} style={style.remove} />
      </TouchableHighlight>
    )
  }

  onPressError = () => {
    ActionSheetIOS.showActionSheetWithOptions({
      title: i18n('Failed to upload attachment'),
      message: this.props.error,
      options: [
        i18n('Retry Upload'),
        i18n('Delete'),
        i18n('Cancel'),
      ],
      destructiveButtonIndex: 1,
      cancelButtonIndex: 2,
    }, (i) => [this.props.onRetry, this.props.onRemovePressed, () => {}][i]())
  }

  onPressRemove = () => {
    Alert.alert(
      i18n('Remove this attachment?'),
      i18n('This action can not be undone.'),
      [
        { text: i18n('Cancel'), onPress: null, style: 'cancel' },
        { text: i18n('Remove'), onPress: this.props.onRemovePressed, style: 'destructive' },
      ],
    )
  }
}

const style = StyleSheet.create({
  remove: {
    width: 14,
    height: 14,
    tintColor: 'black',
  },
  cancel: {
    position: 'absolute',
    top: 0,
    right: 0,
    bottom: 0,
    left: 0,
    alignItems: 'center',
    justifyContent: 'center',
  },
  cancelInner: {
    width: 6,
    height: 6,
  },
  image: {},
})
