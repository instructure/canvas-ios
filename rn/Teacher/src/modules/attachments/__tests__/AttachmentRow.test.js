//
// This file is part of Canvas.
// Copyright (C) 2017-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import React from 'react'
import { Alert } from 'react-native'
import renderer from 'react-test-renderer'
import AttachmentRow, { type Props } from '../AttachmentRow'
import explore from '../../../../test/helpers/explore'

jest
  .mock('react-native/Libraries/Components/Button', () => 'Button')
  .mock('react-native/Libraries/Components/Touchable/TouchableHighlight', () => 'TouchableHighlight')
  .mock('react-native/Libraries/Components/Touchable/TouchableOpacity', () => 'TouchableOpacity')

describe('AttachmentRow', () => {
  let props
  beforeEach(() => {
    props = {
      completed: true,
      title: 'Attachment 1',
      subtitle: 'Uploading...',
      onPress: jest.fn(),
      testID: 'attachment-row.0',
      onRemovePressed: jest.fn(),
      progress: { loaded: 0, total: 0 },
    }
  })

  it('renders', () => {
    expect(render(props).toJSON()).toMatchSnapshot()
  })

  it('prompts to remove', () => {
    props.testID = 'attachment-row'
    const spy = jest.fn()
    // $FlowFixMe
    Alert.alert = spy
    const removeBtn: any = explore(render(props).toJSON()).selectByID('attachment-row.remove.btn')
    removeBtn.props.onPress()
    expect(spy).toHaveBeenCalledWith(
      'Remove this attachment?',
      'This action can not be undone.',
      [
        { text: 'Cancel', onPress: null, style: 'cancel' },
        { text: 'Remove', onPress: expect.any(Function), style: 'destructive' },
      ],
    )
  })

  it('calls onRemovePressed when alert confirms', () => {
    const spy = jest.fn()
    props.onRemovePressed = spy
    props.testID = 'attachment-row'
    // $FlowFixMe
    Alert.alert = jest.fn((title, message, buttons) => buttons[1].onPress())
    const removeBtn: any = explore(render(props).toJSON()).selectByID('attachment-row.remove.btn')
    removeBtn.props.onPress()
    expect(spy).toHaveBeenCalled()
  })

  function render (props: Props): any {
    return renderer.create(<AttachmentRow {...props} />)
  }
})
