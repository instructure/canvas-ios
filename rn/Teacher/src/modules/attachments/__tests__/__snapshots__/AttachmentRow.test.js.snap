// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentRow renders 1`] = `
<TouchableHighlight
  accessibilityRole="button"
  accessible={false}
  onPress={[Function]}
  style={
    Array [
      undefined,
      undefined,
    ]
  }
  testID="attachment-row.0"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "flex": 1,
          "flexDirection": "row",
          "minHeight": 54,
          "paddingBottom": 12,
          "paddingLeft": 16,
          "paddingRight": 16,
          "paddingTop": 12,
        },
        Object {
          "backgroundColor": "#FFFFFF",
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityLabel="Upload complete"
      accessible={true}
      style={Object {}}
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/images/attachments/complete-icon.png",
          }
        }
        testID="attachment-row.0.icon.complete"
      />
    </View>
    <View
      style={
        Array [
          Object {
            "flex": 1,
            "flexDirection": "column",
            "justifyContent": "center",
          },
          Object {
            "marginLeft": 12,
          },
        ]
      }
    >
      <Text
        ellipsizeMode="tail"
        numberOfLines={0}
        style={
          Array [
            Object {
              "color": "#2D3B45",
              "fontFamily": "Lato-Regular",
              "fontSize": 16,
            },
            Array [
              Object {
                "color": "#2D3B45",
                "fontWeight": "600",
              },
            ],
          ]
        }
      >
        Attachment 1
      </Text>
    </View>
    <View
      style={
        Object {
          "flex": 0,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "row",
            "justifyContent": "flex-end",
          }
        }
      >
        <TouchableHighlight
          accessibilityLabel="Remove attachment"
          accessibilityTraits="button"
          hitSlop={
            Object {
              "bottom": 8,
              "left": 8,
              "right": 8,
              "top": 8,
            }
          }
          onPress={[Function]}
          testID="attachment-row.0.remove.btn"
          underlayColor="#FFFFFF"
        >
          <Image
            source={
              Object {
                "testUri": "../../../src/images/x-icon.png",
              }
            }
            style={
              Object {
                "height": 14,
                "tintColor": "#2D3B45",
                "width": 14,
              }
            }
          />
        </TouchableHighlight>
      </View>
    </View>
  </View>
</TouchableHighlight>
`;
